### ISSUE TITLE
prepared statement "$1" does not exist 

### ISSUE THREAD

**Issue by AlexanderInova** [pgadmin-org/pgadmin4#6285]
For a few weeks now, we have been getting the following error message very regularly when running queries in pgAdmin:

```
prepared statement "_pg3_1" does not exist

SQL state: 26000
```

(the numbers in "_pg3_1" might be different sometimes)

Sometimes it occurs in the data grid when saving, sometimes after a few executions of a query and sometimes only after hours. In rare cases the query then works after the 3-5th attempt. In most cases it only helps to close the query and open a new one. Especially when editing data in the DataGrid, this is annoying, because the changes cannot be applied, or only in a cumbersome way. 

Unfortunately, there are currently no known steps to reproduce this issue. It just happens some times when trying to execute queries. 

**Expected behavior**

Queries get executed without raising the described error.

**Error message**

```
prepared statement "$1" does not exist 

SQL state: 26000
```

**Screenshots**

![image](https://github.com/pgadmin-org/pgadmin4/assets/113263664/393d08e1-005a-4808-817c-04528a429c3e)

**Additional context**
pgAdmin is runnig in Azure Kubernetes Services (AKS). 
We are using an "Azure Database for PostgreSQL flexible servers". That's a Postgre server behind a pgBouncer. 

**Note**: None of our other applications raised this error so far.

**Comment by FaharAbbasRizvi** [comment]
@Alexanerinova 

I am not able to reproduce this issue, Can please check the same issue on the latest release of pgadmin4 7.1 and share the exact to reproduce or share the small video?

**Comment by Alexanerinova** [comment]
@FaharAbbasRizvi 

Here is a video, even I don't think you could gather more information from that, than from a screenshot: 
[2023-05-27 20-44-17-censored.webm](https://github.com/pgadmin-org/pgadmin4/assets/113263664/0d6422d7-fa45-445d-84a1-76e36cd5cafd)

The size of the statement does not matter. This also happened with a statement as simple as "SELECT 1;", or when editing a small amount of data with a datagrid. In this case, the error did not disappear just by repeating the execution. I had to open a new query. In rare cases, the execution works after repeating the execution a few times.

**Comment by Alexanerinova** [comment]
Still reproducible with Version 7.3.

**Comment by FaharAbbasRizvi** [comment]
@Alexanerinova 

I am still not able to reproduce the issue. 

Could you please try to execute the same command from PSQL(Command line tools for PostgreSQL) and share the results?

**Comment by Alexanerinova** [comment]
@FaharAbbasRizvi The issue is most probably not related to the text of the command. It appears completely random and in rare cases disappears by just trying to repeat the execution multiple times. According to our centralized log, none of our applications ever run into this issue, so I think it is also not likely to be related to the PostgreSQL oder pgBouncer in general. However, we do not use prepared statements, which seem to be the issue here. 

If it helps, I will try to do some of my future work with psql and check if the issue appears this way, too. 

I don't know how exactly pgAdmin4 works under the hood, but, according to your request, it uses the psql command line util for that? Please confirm, if this is the case. If so, according to the error message, the tool seems to prepare a statement, which then gets lost somehow. In this case it might be an issue between the npgsql prepared statement, pgBouncer and PostgreSQL.

**Comment by FaharAbbasRizvi** [comment]
@Alexanerinova 

I have tested the same behaviour with multi-users with google chrome, firefox and Edge browser and connected with pgbouncer with wheel package. But I am not able to see such kinds of issues.

Could you please elaborate more,  "Steps to reproduce", "Installation mode", and "how you connected with pgbouncer"?

<img width="1142" alt="Screen Shot 2023-06-15 at 5 19 50 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/43006122/f7f6c733-db23-4f03-bcd0-b9c8aab674fe">

**Comment by Alexanerinova** [comment]
@FaharAbbasRizvi 
Thank you for your investigation so far. Even in our environment this happens only every now and then. 

Here our setup: 
- pgAdmin is running as a container/deployment in a AKS (Azure Kubernetes Services) Cluster ([Manifest](https://github.com/pgadmin-org/pgadmin4/files/11771181/pgadmin-manifest-censored.zip))
- istio is used as service mesh and ingress controller. Anyway istio envoy proxies are disabled for the pgAdmin4, so its egress controller is bypassed for accessing pgBouncer 
- As database server we use a "Azure Database for PostgreSQL - Flexible Server" (version 13.10 - we cannot upgrade since Microsoft does not support inplace-updates, but only migrating into a new instance, which is also not possible since the data center is overloaded and thus creating new instances is not allowed) 
- pgBouncer is enabled via "Server Properties" of the Azure Database, running at port 6432 
- As authentication source we use a oauth2/openid provider (Microsoft Azure Active Directory) 
- pgAdmin4s "ENHANCED_COOKIE_PROTECTION" is disabled, since this feature is quite broken when running behind a reverse-proxy / ingress controller
- We are accessing pgAdmin4 over a VPN, which probably doesn't affect the connection between pgAdmin4 and pgBouncer

This issue first appeared a few months ago. Before that, we didn't ever experience this error. We have no idea when exactly the issue appear. Currently, we assume it happens more frequently, if multiple users are concurrently using pgAdmin4, and in rare situations, when a single user works with multiple queries / tabs. While sometimes the issue does not occur for days, it sometimes drives us crazy when it then happens every few minutes or hours on a single day. 

A few examples when it happened: 
1. We executed a pretty simple statement for selecting full rows of a medium size table repeatedly, since we were observing changes made to the data by our cluster. Since we were in a call, the browser window (currently using Microsoft Edge and Mozilla Firefox) might be in the background for some time. Then the issue appeared. Re-executing the query for at least 10-15 times did not fix the issue. The numbers within the error message `prepared statement "_pg3_1" does not exist` counted up with each try. At the same time, executing the same statement in a new query/tab worked fine. 
2. We tested the statement for some calculations without using any database tables at all (something like "SELECT 1+2;", for a quite simplified example). After some time the error occurred. The number within the error message did not change when repeating the statement. 
3. On another day we worked with bigger sets of data and quite complex statements. After working a while, the error occurred. After retrying the statement 2 or 3 times, the error disappeared and the query executed successfully. 
4. At one night, when nobody else was working, we executed a quite simple statement and ran immediately into this issue (which was quite surprising - up to then we were quite convinced, that the issue only appears when using a connection for a longer period of time). 

After reading that, do you have an idea what might cause the issue or what we could try to reproduce?

I will instruct my colleagues to bypass the pgBouncer for the next few weeks in order to see if the issue might be related to pgBouncer. But since we don't know how to reproduce the issue reliably, we can't say if it's actually gone, even after testing for weeks.

**Comment by NTMan** [comment]
I have same issue. It started after 7.0 version and not fixed yet in 7.4 version. It definitely happens after executing several sql queries (count of queries is varying). And for workaround this I disconnect/and connect again to desired database.

**Comment by DnikitaA** [comment]
Psycopg uses an automatic system to manage prepared statements. A query is prepared automatically after it is executed more than prepare_threshold.Connection.prepare_threshold) times on a connection. By default prepare_threshold = 5.

For reproduce the bug, you need to configure the following environment:
- Postgresql (for example v15) with default configuration.
- Create two databases on the postgreslq server: test1 and test2
- PgBouncer (for example v1.16.1).
- Override statements in pgbouncer config:
```
[databases]
* = host=localhost port=5432 pool_size=5
[users]
postgres = max_user_connections=1
[pgbouncer]
listen_addr = *
listen_port = 6432
auth_type = md5
auth_file = /etc/pgbouncer/userlist.txt
```
- Override userlist
```
"postgres" "postgres"
```
- Pgadmin4 above version 7 (For example v7.3)
- Connect server over port 6432 (over pgbouncer)
- open two tabs with query editor for test1 and test2 databases
- execute query ```select 1``` for test1 database more than 5 times (min 6)
- switch tab and execute query ```select 1``` for test2 database
- switch tab and execute query ```select 1``` for test1 database

**Comment by FaharAbbasRizvi** [comment]
@DnikitaA 

I am getting a different  message "connection timeout error" If I follow the steps you have already shared with pgbouncer 

<img width="1317" alt="Screen Shot 2023-08-28 at 4 32 37 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/43006122/a7385523-649c-4c57-b258-4654cdc3cf4a">


Please try the same steps without pgbouncer if not then this could be an issue of pgbouncer.

I have  already reported a similar issue with pgbouncer https://github.com/pgadmin-org/pgadmin4/issues/6304

Could you please verify same issue on the last release of 7.6?

**Comment by NTMan** [comment]
Looks like I bisect the issue

```

**Comment by NTMan** [comment]
@khushboovashi ^^^

**Comment by wibrt** [comment]
Hello,

it seems we might be experiencing the same issue.

Could it be that pgadmin expects connections to be session connections, whereas a pooling server might just serve statement or transaction pooling mode, leading to these problems.

I can understand that for administrative queries are prepared for performance,
but this wil give problems when there is no session connection

Concrete use case: pgadmin>pgboucer>postgresql

I see prep statements in the pglogs like,( usually in the form of `_pg3_*`):

```
LOG:  duration: 0.155 ms  bind _pg3_0: SELECT at.attname, ty.typname, at.attnum
5729476-	    FROM pg_catalog.pg_attribute at
5729477-	    LEFT JOIN pg_catalog.pg_type ty ON (ty.oid = at.atttypid)
5729478-	    JOIN pg_catalog.pg_class as cl ON cl.oid=at.attrelid
5729479-	    JOIN pg_catalog.pg_namespace as nsp ON nsp.oid=cl.relnamespace
5729480-	WHERE
5729481-	    attrelid=31893::oid AND
5729482-	at.attnum > 0
5729483-	AND at.attisdropped = FALSE
```

but also

`ERROR:  prepared statement "_pg3_0" does not exist`

The end users complain about the same error, ie  "_pg3_0" ..

a solution: could there be an option in pgadmin to disable prepared statements at a server connection level (register server in pgadmin)

similar to prepareThreshold=0 in java,
https://www.pgbouncer.org/faq.html

hth

**Comment by adityatoshniwal** [comment]
Thank you @wibrt for your valuable inputs. I have created a PR - https://github.com/pgadmin-org/pgadmin4/pull/6853 which allow setting prepare threshold in server connection.

**Comment by nikhil-mohite** [comment]
Resolved and tested on latest snapshot build: https://www.postgresql.org/ftp/pgadmin/pgadmin4/snapshots/2023-10-16/
OS: macOS (Ventura 13.3)
Mode: Desktop

**Comment by MHenn1g** [comment]
Thanks @nikhil-mohite, 
maybe I'm wrong about these images, however I tried utilizing dpage/pgadmin4:2023-10-16-1 as well as dpage/pgadmin4:2023-10-18-2 in order to get rid of this error, however they still appear.

Is there something you need to configure on the server side in order to remove the bug?

Best regards :)

**Comment by AlexanderInova** [comment]
We tested this workaround for a while and it seems to work fine so far. Thanks to everyone involved.

**Comment by davd-edia** [comment]
Hello. I'm seeing this behavior on pgadmin 8.1.

**Comment by adityatoshniwal** [comment]
> Hello. I'm seeing this behavior on pgadmin 8.1.

Check the `Prepare threshold` field in https://www.pgadmin.org/docs/pgadmin4/8.1/server_dialog.html

**Comment by davd-edia** [comment]
Thanks! I've set to 0, but now I can't connect to the server, a new error occurs when I submit the password:
"cannot insert multiple commands into a prepared statement"
